src/handler.rs

```diff
@@ -3,14 +3,78 @@ use std::future::Future;
 use std::pin::Pin;
 use std::sync::Arc;
 
-use futures::channel::oneshot::Sender as SyncSender;
+use futures::channel::oneshot::{self, Sender as SyncSender};
+use futures::FutureExt;
 
 use crate::actor::{Actor, AsyncContext};
 use crate::address::Addr;
 use crate::context::Context;
-use crate::fut::ActorFuture;
+use crate::fut::{wrap_future, ActorFuture};
 use crate::WrapFuture;
 
+pub struct TempRef<T> {
+    inner: *mut T,
+}
+
+unsafe impl<T: Send> Send for TempRef<T> {}
+
+impl<T> TempRef<T> {
+    pub fn as_ref(&self) -> &T {
+        unsafe {
+            let x = self.inner;
+            &*x
+        }
+    }
+    pub fn as_mut(&mut self) -> &mut T {
+        unsafe {
+            let x = self.inner;
+            &mut *x
+        }
+    }
+}
+
+pub trait AsyncHandler<M>
+where
+    Self: Actor,
+    <Self as Actor>::Context: AsyncContext<Self>,
+    M: AsyncMessage,
+{
+    type Result: Future<Output = M::Result> + Sized + 'static;
+
+    fn handle(
+        actor: TempRef<Self>,
+        msg: M,
+        ctx: TempRef<Self::Context>,
+    ) -> Self::Result;
+}
+
+impl<M, I: Send + 'static, AH> Handler<M> for AH
+where
+    AH: AsyncHandler<M> + 'static,
+    <AH as Actor>::Context: AsyncContext<AH>,
+    M: AsyncMessage<Result = I> + 'static,
+{
+    type Result = ResponseFuture<AsyncResponse<M>>;
+
+    fn handle(&mut self, msg: M, ctx: &mut Self::Context) -> Self::Result {
+        let actor = TempRef { inner: self };
+
+        let tmp_ctx = TempRef { inner: ctx };
+
+        let fut_res = Self::handle(actor, msg, tmp_ctx);
+
+        let (s, r) = oneshot::channel::<M::Result>();
+
+        ctx.wait(wrap_future(async move {
+            let res = fut_res.await;
+            s.send(res)
+                .unwrap_or_else(|_| panic!("Failed to send future result"));
+        }));
```

Comment:
Here we are blocking processing of messages while future completes.
What also could be done is awaiting for `TempRef` of actor and context to be dropped. This way we can ensure that even if those reverences were sent to some other threads, we still won't have memory issues.